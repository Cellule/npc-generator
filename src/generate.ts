import type { DebugNode, GenerateOptionValueSimple, Npc, NpcGenerateOptions, NpcGenerateOptionsValues, NpcGenerateResult } from "../index.d";
import { getGroups } from "./groups";
import schema from "./schema.json";
import { getNamedTableOptions, getTableReferenceOptions } from "./tables";
import { Group, Operator, SchemaDescriptor, SchemaElement, SchemaResult, WeightedValue } from "./types.js";
import { chooseRandomWithWeight } from "./utils";

function numberOrNull(v: any) {
  return typeof v === "number" ? v | 0 : null;
}

export function generate({ npcOptions }: { npcOptions?: NpcGenerateOptions } = {}): NpcGenerateResult {
  const options: Required<NpcGenerateOptions> = {
    race: numberOrNull(npcOptions?.race),
    subrace: numberOrNull(npcOptions?.subrace),
    classorprof: numberOrNull(npcOptions?.classorprof),
    occupation1: numberOrNull(npcOptions?.occupation1),
    occupation2: numberOrNull(npcOptions?.occupation2),
    alignment: numberOrNull(npcOptions?.alignment),
    plothook: numberOrNull(npcOptions?.plothook),
    gender: numberOrNull(npcOptions?.gender),
  };

  const context = { vars: {} };
  let debugNode: DebugNode = { o: "root", childs: [] };
  function processGroups(groups: Group[]) {
    let result = "";
    for (const instruction of groups) {
      if (typeof instruction === "string") {
        debugNode.childs.push(instruction);
        result += String(instruction);
      } else {
        const oldNode = debugNode;
        const node = { o: instruction.original, childs: [] };
        debugNode.childs.push(node);
        debugNode = node;
        if (typeof instruction === "function") {
          const insRes = (instruction as Operator)(context, options);
          if (insRes !== undefined) {
            if (Array.isArray(insRes)) {
              result += String(processGroups(insRes));
            } else {
              result += String(insRes);
            }
          }
        } else if (Array.isArray(instruction)) {
          result += String(processGroups(instruction));
        }
        debugNode = oldNode;
      }
    }
    return result;
  }

  function chooseFromArray(arr: WeightedValue[]): string {
    const totalWeight = arr.reduce(function (w, e) {
      return w + (e.w | 0);
    }, 0);
    return chooseRandomWithWeight(arr, totalWeight);
  }

  function processSchema(schemaElement: SchemaDescriptor): SchemaResult;
  function processSchema(schemaElement: any): string;
  function processSchema(schemaElement: SchemaElement | SchemaDescriptor): SchemaResult | string {
    if (typeof schemaElement === "string") {
      return processGroups(getGroups(schemaElement));
    } else if (Array.isArray(schemaElement)) {
      return processSchema(chooseFromArray(schemaElement));
    }
    const result: SchemaResult = {};
    for (const name of Object.keys(schemaElement)) {
      const element = schemaElement[name];
      // need to make a choice based on weight
      result[name] = processSchema(element);
    }
    return result;
  }

  // process inititalisation first, most of the selection is done here
  processGroups(getGroups(schema.options.initialisation));

  // Force to Npc type because we know schema.output matches the Npc type
  const npc: Npc = processSchema(schema.output) as any;
  return { npc, debugNode };
}

export function getNpcOptionsValues(): NpcGenerateOptionsValues {
  const subraces: Record<string, GenerateOptionValueSimple | undefined> = {
    elf: getNamedTableOptions("raceelf"),
    dwarf: getNamedTableOptions("racedwarf"),
    gnome: getNamedTableOptions("racegnome"),
    halfling: getNamedTableOptions("racehalfling"),
  };
  const races = getTableReferenceOptions("race").map(({ name, table }) => ({
    name,
    subraces: subraces[table],
  }));

  const genders = getNamedTableOptions("gender");
  const alignments = getNamedTableOptions("forcealign");
  const plothooks = getNamedTableOptions("hooks");
  const classes = getNamedTableOptions("class");

  const professionCategories: Record<string, GenerateOptionValueSimple | undefined> = {
    learned: getNamedTableOptions("learned"),
    lesserNobility: getNamedTableOptions("lesserNobility"),
    professional: getNamedTableOptions("professional"),
    workClass: getNamedTableOptions("workClass"),
    martial: getNamedTableOptions("martial"),
    underclass: getNamedTableOptions("underclass"),
    entertainer: getNamedTableOptions("entertainer"),
  };
  const professions = getTableReferenceOptions("profession").map(({ name, table }) => ({
    name,
    professionCategories: professionCategories[table],
  }));

  return {
    races,
    genders,
    alignments,
    plothooks,
    classes,
    professions,
  };
}
