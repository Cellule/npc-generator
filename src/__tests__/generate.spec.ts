import { debugNodeToString } from "../..";
import { generate, getNpcOptionsValues } from "../generate";

describe("generate", () => {
  it("generates a random npc that respects the schema/type", () => {
    const { npc } = generate();

    const { abilities, alignment, description, hook, physical, pquirks, ptraits, relationship, religion } = npc;
    // check abilities are positive numbers
    expect(abilities).toEqual(
      expect.objectContaining({
        charisma: expect.any(Number),
        constitution: expect.any(Number),
        dexterity: expect.any(Number),
        intelligence: expect.any(Number),
        strength: expect.any(Number),
        wisdom: expect.any(Number),
      }),
    );
    // check alignment are positive numbers
    expect(alignment).toEqual(
      expect.objectContaining({
        good: expect.any(Number),
        moralneutral: expect.any(Number),
        evil: expect.any(Number),
        lawful: expect.any(Number),
        ethicalneutral: expect.any(Number),
        chaotic: expect.any(Number),
      }),
    );

    // check descriptions
    expect(description).toEqual(
      expect.objectContaining({
        name: expect.any(String),
        age: expect.any(Number),
        kenkuname: expect.any(String),
        gender: expect.any(String),
        race: expect.any(String),
        occupation: expect.any(String),
        pronounMinus: expect.any(String),
        pronounCapit: expect.any(String),
      }),
    );

    // check physical
    expect(physical).toEqual(
      expect.objectContaining({
        hair: expect.any(String),
        eyes: expect.any(String),
        skin: expect.any(String),
        height: expect.any(Number),
        build: expect.any(String),
        face: expect.any(String),
        special1: expect.any(String),
        special2: expect.any(String),
      }),
    );

    // check relationship
    expect(relationship).toEqual(
      expect.objectContaining({
        orientation: expect.any(String),
        status: expect.any(String),
      }),
    );

    // check religion
    expect(religion).toEqual(
      expect.objectContaining({
        description: expect.any(String),
      }),
    );

    // check ptraits
    expect(ptraits).toEqual(
      expect.objectContaining({
        traitslizards: expect.any(String),
        traitsgoliaths: expect.any(String),
        traits1: expect.any(String),
        traits2: expect.any(String),
      }),
    );

    // check pquirks
    expect(pquirks).toEqual(
      expect.objectContaining({
        description: expect.any(String),
      }),
    );

    // check hook
    expect(hook).toEqual(
      expect.objectContaining({
        description: expect.any(String),
      }),
    );
  });

  const options = getNpcOptionsValues();
  it("returns expected npc options", () => {
    expect(options).toMatchSnapshot("npc-options");
  });

  it.each(options.races)("always generates expected race $name", (race) => {
    const { npc } = generate({
      npcOptions: { race: race.value },
    });
    const raceLc = race.name.toLowerCase();
    const generatedRaceLc = npc.description.race.toLowerCase();
    if (raceLc === "lizardfolk") {
      expect(generatedRaceLc).toContain("lizard");
    } else {
      if (race.subraces?.length) {
        expect(generatedRaceLc).toMatch(new RegExp(race.subraces.map((subrace) => subrace.name.toLowerCase()).join("|")));
      } else {
        expect(generatedRaceLc).toContain(raceLc);
      }
    }
  });

  const subRacesCases = options.races.flatMap((race) => race.subraces?.map((subrace) => ({ race, subrace })));
  it.each(subRacesCases.filter((o): o is Exclude<typeof o, undefined> => !!o))(
    "always generates expected subrace $race.name:$subrace.name",
    ({ race, subrace }) => {
      const { npc } = generate({
        npcOptions: { race: race.value, subrace: subrace.value },
      });
      const subraceLc = subrace.name.toLowerCase();
      const generatedRaceLc = npc.description.race.toLowerCase();
      expect(generatedRaceLc).toContain(subraceLc);
    },
  );

  it.each(options.genders)("always generates expected gender $name", (gender) => {
    const { npc } = generate({
      npcOptions: {
        gender: gender.value,
      },
    });

    const genderLc = gender.name.toLowerCase();
    const generatedGenderLc = npc.description.gender.toLowerCase();
    expect(generatedGenderLc).toContain(genderLc);
  });

  it.each(options.alignments)("always generates expected alignment $name", (alignment) => {
    const { npc } = generate({
      npcOptions: {
        alignment: alignment.value,
      },
    });
    if (alignment.name === "Good") {
      expect(npc.alignment.good).toBeGreaterThan(npc.alignment.evil);
    } else {
      expect(npc.alignment.evil).toBeGreaterThan(npc.alignment.good);
    }
  });

  it.each(options.classes)("always generates expected class $name", (class_) => {
    const { npc } = generate({
      npcOptions: {
        classorprof: 0,
        occupation1: class_.value,
      },
    });

    const classLc = class_.name.toLowerCase();
    const generatedClassLc = npc.description.occupation.toLowerCase();
    expect(generatedClassLc).toContain(classLc);
  });

  it.each(options.plothooks)("always generates expected plothook $name", (plothook) => {
    const { npc, debugNode } = generate({
      npcOptions: {
        plothook: plothook.value,
      },
    });
    const path = debugNodeToString(debugNode);
    expect(path).toContain(plothook.name === "Classic" ? `-> {hooks1}` : `-> {hooks2}`);
  });

  // TODO:: test Professions options

  it.each(options.professions)("profession $name should have subcategories", (profession) => {
    expect(profession.professionCategories?.length).toBeGreaterThan(1);
  });
});
