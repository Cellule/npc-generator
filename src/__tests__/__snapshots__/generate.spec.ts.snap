// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate returns expected npc options: npc-options 1`] = `
{
  "alignments": [
    {
      "name": "Good",
      "value": 1,
    },
    {
      "name": "Evil",
      "value": 2,
    },
  ],
  "classes": [
    {
      "name": "Artificier",
      "value": 0,
    },
    {
      "name": "Barbarian",
      "value": 1,
    },
    {
      "name": "Bard",
      "value": 2,
    },
    {
      "name": "Cleric",
      "value": 3,
    },
    {
      "name": "Druid",
      "value": 4,
    },
    {
      "name": "Fighter",
      "value": 5,
    },
    {
      "name": "Monk",
      "value": 6,
    },
    {
      "name": "Mystic",
      "value": 7,
    },
    {
      "name": "Paladin",
      "value": 8,
    },
    {
      "name": "Ranger",
      "value": 9,
    },
    {
      "name": "Rogue",
      "value": 10,
    },
    {
      "name": "Sorcerer",
      "value": 11,
    },
    {
      "name": "Warlock",
      "value": 12,
    },
    {
      "name": "Wizard",
      "value": 13,
    },
  ],
  "genders": [
    {
      "name": "Male",
      "value": 0,
    },
    {
      "name": "Female",
      "value": 1,
    },
  ],
  "occupations": [
    {
      "classes": [
        {
          "name": "Artificier",
          "value": 0,
        },
        {
          "name": "Barbarian",
          "value": 1,
        },
        {
          "name": "Bard",
          "value": 2,
        },
        {
          "name": "Cleric",
          "value": 3,
        },
        {
          "name": "Druid",
          "value": 4,
        },
        {
          "name": "Fighter",
          "value": 5,
        },
        {
          "name": "Monk",
          "value": 6,
        },
        {
          "name": "Mystic",
          "value": 7,
        },
        {
          "name": "Paladin",
          "value": 8,
        },
        {
          "name": "Ranger",
          "value": 9,
        },
        {
          "name": "Rogue",
          "value": 10,
        },
        {
          "name": "Sorcerer",
          "value": 11,
        },
        {
          "name": "Warlock",
          "value": 12,
        },
        {
          "name": "Wizard",
          "value": 13,
        },
      ],
      "name": "Class",
      "professions": undefined,
      "value": 0,
    },
    {
      "classes": undefined,
      "name": "Profession",
      "professions": [
        {
          "name": "Learned",
          "professionCategories": [
            {
              "name": "Astrologer",
              "value": 0,
            },
            {
              "name": "Cartographer",
              "value": 1,
            },
            {
              "name": "Historian",
              "value": 2,
            },
            {
              "name": "Poet",
              "value": 3,
            },
            {
              "name": "Philosopher",
              "value": 4,
            },
            {
              "name": "Hermit",
              "value": 5,
            },
            {
              "name": "Wandering Pilgrim",
              "value": 6,
            },
            {
              "name": "Barber",
              "value": 7,
            },
            {
              "name": "Doctor",
              "value": 8,
            },
            {
              "name": "Medic",
              "value": 9,
            },
            {
              "name": "Barrister",
              "value": 10,
            },
            {
              "name": "Herald",
              "value": 11,
            },
            {
              "name": "Scribe",
              "value": 12,
            },
            {
              "name": "Acolyte",
              "value": 13,
            },
            {
              "name": "Neophyte",
              "value": 14,
            },
          ],
          "value": 0,
        },
        {
          "name": "Lesser Nobility",
          "professionCategories": [
            {
              "name": "Explorer",
              "value": 0,
            },
            {
              "name": "Diplomat",
              "value": 1,
            },
            {
              "name": "Knight",
              "value": 2,
            },
            {
              "name": "Minister",
              "value": 3,
            },
            {
              "name": "Page",
              "value": 4,
            },
            {
              "name": "Squire",
              "value": 5,
            },
          ],
          "value": 1,
        },
        {
          "name": "Professional",
          "professionCategories": [
            {
              "name": "Armorer",
              "value": 0,
            },
            {
              "name": "Painter",
              "value": 1,
            },
            {
              "name": "Baker",
              "value": 2,
            },
            {
              "name": "Banker",
              "value": 3,
            },
            {
              "name": "Blacksmith",
              "value": 4,
            },
            {
              "name": "Bowyer",
              "value": 5,
            },
            {
              "name": "Brewer",
              "value": 6,
            },
            {
              "name": "Butcher",
              "value": 7,
            },
            {
              "name": "Carpenter",
              "value": 8,
            },
            {
              "name": "Clothier",
              "value": 9,
            },
            {
              "name": "Cook",
              "value": 10,
            },
            {
              "name": "Furrier",
              "value": 11,
            },
            {
              "name": "Goldsmith",
              "value": 12,
            },
            {
              "name": "Innkeeper",
              "value": 13,
            },
            {
              "name": "Jeweler",
              "value": 14,
            },
            {
              "name": "Leatherworker",
              "value": 15,
            },
            {
              "name": "Locksmith",
              "value": 16,
            },
            {
              "name": "Mason",
              "value": 17,
            },
            {
              "name": "Merchant",
              "value": 18,
            },
            {
              "name": "Musician",
              "value": 19,
            },
            {
              "name": "Tinker",
              "value": 20,
            },
            {
              "name": "Torturer",
              "value": 21,
            },
            {
              "name": "Trader",
              "value": 22,
            },
            {
              "name": "Weaver",
              "value": 23,
            },
          ],
          "value": 2,
        },
        {
          "name": "Working Class",
          "professionCategories": [
            {
              "name": "Boat",
              "value": 0,
            },
            {
              "name": "Coach",
              "value": 1,
            },
            {
              "name": "Farmer",
              "value": 2,
            },
            {
              "name": "Fisher",
              "value": 3,
            },
            {
              "name": "Gravedigger",
              "value": 4,
            },
            {
              "name": "Sheppard",
              "value": 5,
            },
            {
              "name": "Trapper",
              "value": 6,
            },
            {
              "name": "Messenger",
              "value": 7,
            },
            {
              "name": "Miller",
              "value": 8,
            },
            {
              "name": "Miner",
              "value": 9,
            },
            {
              "name": "Peddler",
              "value": 10,
            },
            {
              "name": "Ratcatcher",
              "value": 11,
            },
            {
              "name": "Sailor",
              "value": 12,
            },
          ],
          "value": 3,
        },
        {
          "name": "Martial",
          "professionCategories": [
            {
              "name": "Bodyguard",
              "value": 0,
            },
            {
              "name": "Bounty Hunter",
              "value": 1,
            },
            {
              "name": "Forester",
              "value": 2,
            },
            {
              "name": "Gladiator",
              "value": 3,
            },
            {
              "name": "Jailer",
              "value": 4,
            },
            {
              "name": "Soldier",
              "value": 5,
            },
          ],
          "value": 4,
        },
        {
          "name": "Underclass",
          "professionCategories": [
            {
              "name": "Bandit",
              "value": 0,
            },
            {
              "name": "Beggar",
              "value": 1,
            },
            {
              "name": "Fence",
              "value": 2,
            },
            {
              "name": "Pickpocket",
              "value": 3,
            },
            {
              "name": "Procurer",
              "value": 4,
            },
            {
              "name": "Prostitute",
              "value": 5,
            },
            {
              "name": "Slaver",
              "value": 6,
            },
            {
              "name": "Smuggler",
              "value": 7,
            },
          ],
          "value": 5,
        },
        {
          "name": "Entertainer",
          "professionCategories": [
            {
              "name": "Acrobat",
              "value": 0,
            },
            {
              "name": "Actor",
              "value": 1,
            },
            {
              "name": "Clown",
              "value": 2,
            },
            {
              "name": "Dancer",
              "value": 3,
            },
            {
              "name": "Fortune-Teller",
              "value": 4,
            },
            {
              "name": "Juggler",
              "value": 5,
            },
            {
              "name": "Prestidigitator",
              "value": 6,
            },
            {
              "name": "Funambulist",
              "value": 7,
            },
            {
              "name": "Storyteller",
              "value": 8,
            },
          ],
          "value": 6,
        },
      ],
      "value": 1,
    },
  ],
  "plothooks": [
    {
      "name": "Classic",
      "value": 0,
    },
    {
      "name": "Funky",
      "value": 1,
    },
  ],
  "professions": [
    {
      "name": "Learned",
      "professionCategories": [
        {
          "name": "Astrologer",
          "value": 0,
        },
        {
          "name": "Cartographer",
          "value": 1,
        },
        {
          "name": "Historian",
          "value": 2,
        },
        {
          "name": "Poet",
          "value": 3,
        },
        {
          "name": "Philosopher",
          "value": 4,
        },
        {
          "name": "Hermit",
          "value": 5,
        },
        {
          "name": "Wandering Pilgrim",
          "value": 6,
        },
        {
          "name": "Barber",
          "value": 7,
        },
        {
          "name": "Doctor",
          "value": 8,
        },
        {
          "name": "Medic",
          "value": 9,
        },
        {
          "name": "Barrister",
          "value": 10,
        },
        {
          "name": "Herald",
          "value": 11,
        },
        {
          "name": "Scribe",
          "value": 12,
        },
        {
          "name": "Acolyte",
          "value": 13,
        },
        {
          "name": "Neophyte",
          "value": 14,
        },
      ],
      "value": 0,
    },
    {
      "name": "Lesser Nobility",
      "professionCategories": [
        {
          "name": "Explorer",
          "value": 0,
        },
        {
          "name": "Diplomat",
          "value": 1,
        },
        {
          "name": "Knight",
          "value": 2,
        },
        {
          "name": "Minister",
          "value": 3,
        },
        {
          "name": "Page",
          "value": 4,
        },
        {
          "name": "Squire",
          "value": 5,
        },
      ],
      "value": 1,
    },
    {
      "name": "Professional",
      "professionCategories": [
        {
          "name": "Armorer",
          "value": 0,
        },
        {
          "name": "Painter",
          "value": 1,
        },
        {
          "name": "Baker",
          "value": 2,
        },
        {
          "name": "Banker",
          "value": 3,
        },
        {
          "name": "Blacksmith",
          "value": 4,
        },
        {
          "name": "Bowyer",
          "value": 5,
        },
        {
          "name": "Brewer",
          "value": 6,
        },
        {
          "name": "Butcher",
          "value": 7,
        },
        {
          "name": "Carpenter",
          "value": 8,
        },
        {
          "name": "Clothier",
          "value": 9,
        },
        {
          "name": "Cook",
          "value": 10,
        },
        {
          "name": "Furrier",
          "value": 11,
        },
        {
          "name": "Goldsmith",
          "value": 12,
        },
        {
          "name": "Innkeeper",
          "value": 13,
        },
        {
          "name": "Jeweler",
          "value": 14,
        },
        {
          "name": "Leatherworker",
          "value": 15,
        },
        {
          "name": "Locksmith",
          "value": 16,
        },
        {
          "name": "Mason",
          "value": 17,
        },
        {
          "name": "Merchant",
          "value": 18,
        },
        {
          "name": "Musician",
          "value": 19,
        },
        {
          "name": "Tinker",
          "value": 20,
        },
        {
          "name": "Torturer",
          "value": 21,
        },
        {
          "name": "Trader",
          "value": 22,
        },
        {
          "name": "Weaver",
          "value": 23,
        },
      ],
      "value": 2,
    },
    {
      "name": "Working Class",
      "professionCategories": [
        {
          "name": "Boat",
          "value": 0,
        },
        {
          "name": "Coach",
          "value": 1,
        },
        {
          "name": "Farmer",
          "value": 2,
        },
        {
          "name": "Fisher",
          "value": 3,
        },
        {
          "name": "Gravedigger",
          "value": 4,
        },
        {
          "name": "Sheppard",
          "value": 5,
        },
        {
          "name": "Trapper",
          "value": 6,
        },
        {
          "name": "Messenger",
          "value": 7,
        },
        {
          "name": "Miller",
          "value": 8,
        },
        {
          "name": "Miner",
          "value": 9,
        },
        {
          "name": "Peddler",
          "value": 10,
        },
        {
          "name": "Ratcatcher",
          "value": 11,
        },
        {
          "name": "Sailor",
          "value": 12,
        },
      ],
      "value": 3,
    },
    {
      "name": "Martial",
      "professionCategories": [
        {
          "name": "Bodyguard",
          "value": 0,
        },
        {
          "name": "Bounty Hunter",
          "value": 1,
        },
        {
          "name": "Forester",
          "value": 2,
        },
        {
          "name": "Gladiator",
          "value": 3,
        },
        {
          "name": "Jailer",
          "value": 4,
        },
        {
          "name": "Soldier",
          "value": 5,
        },
      ],
      "value": 4,
    },
    {
      "name": "Underclass",
      "professionCategories": [
        {
          "name": "Bandit",
          "value": 0,
        },
        {
          "name": "Beggar",
          "value": 1,
        },
        {
          "name": "Fence",
          "value": 2,
        },
        {
          "name": "Pickpocket",
          "value": 3,
        },
        {
          "name": "Procurer",
          "value": 4,
        },
        {
          "name": "Prostitute",
          "value": 5,
        },
        {
          "name": "Slaver",
          "value": 6,
        },
        {
          "name": "Smuggler",
          "value": 7,
        },
      ],
      "value": 5,
    },
    {
      "name": "Entertainer",
      "professionCategories": [
        {
          "name": "Acrobat",
          "value": 0,
        },
        {
          "name": "Actor",
          "value": 1,
        },
        {
          "name": "Clown",
          "value": 2,
        },
        {
          "name": "Dancer",
          "value": 3,
        },
        {
          "name": "Fortune-Teller",
          "value": 4,
        },
        {
          "name": "Juggler",
          "value": 5,
        },
        {
          "name": "Prestidigitator",
          "value": 6,
        },
        {
          "name": "Funambulist",
          "value": 7,
        },
        {
          "name": "Storyteller",
          "value": 8,
        },
      ],
      "value": 6,
    },
  ],
  "races": [
    {
      "name": "Aasimar",
      "subraces": undefined,
      "value": 0,
    },
    {
      "name": "Dragonborn",
      "subraces": undefined,
      "value": 1,
    },
    {
      "name": "Dwarf",
      "subraces": [
        {
          "name": "Mountain Dwarf",
          "value": 0,
        },
        {
          "name": "Hill Dwarf",
          "value": 1,
        },
      ],
      "value": 2,
    },
    {
      "name": "Elf",
      "subraces": [
        {
          "name": "Drow",
          "value": 0,
        },
        {
          "name": "High Elf",
          "value": 1,
        },
        {
          "name": "Wood Elf",
          "value": 2,
        },
      ],
      "value": 3,
    },
    {
      "name": "Firbolg",
      "subraces": undefined,
      "value": 4,
    },
    {
      "name": "Gnome",
      "subraces": [
        {
          "name": "Forest Gnome",
          "value": 0,
        },
        {
          "name": "Rock Gnome",
          "value": 1,
        },
      ],
      "value": 5,
    },
    {
      "name": "Goblin",
      "subraces": undefined,
      "value": 6,
    },
    {
      "name": "Goliath",
      "subraces": undefined,
      "value": 7,
    },
    {
      "name": "Halfling",
      "subraces": [
        {
          "name": "Lightfoot Halfling",
          "value": 0,
        },
        {
          "name": "Stout Halfling",
          "value": 1,
        },
      ],
      "value": 8,
    },
    {
      "name": "Half-Elf",
      "subraces": undefined,
      "value": 9,
    },
    {
      "name": "Half-Orc",
      "subraces": undefined,
      "value": 10,
    },
    {
      "name": "Human",
      "subraces": undefined,
      "value": 11,
    },
    {
      "name": "Kenku",
      "subraces": undefined,
      "value": 12,
    },
    {
      "name": "Lizardfolk",
      "subraces": undefined,
      "value": 13,
    },
    {
      "name": "Medusa",
      "subraces": undefined,
      "value": 14,
    },
    {
      "name": "Orc",
      "subraces": undefined,
      "value": 15,
    },
    {
      "name": "Tabaxi",
      "subraces": undefined,
      "value": 16,
    },
    {
      "name": "Tiefling",
      "subraces": undefined,
      "value": 17,
    },
    {
      "name": "Triton",
      "subraces": undefined,
      "value": 18,
    },
    {
      "name": "Troglodyte",
      "subraces": undefined,
      "value": 19,
    },
  ],
}
`;
